On replacing Capybara-webkit (#156)
fastruby/blog: _posts/2020-05-07-removing-capybara-webkit.md

---
layout: post
title: "Removing capybara-webkit"
date: 2020-04-15 12:00:00
categories: ["rails", "upgrades"]
author: arielj
---

We all know testing is important. We have our unit tests and integration tests to make sure everything is working.

> In Ombulabs, we use [capybara](https://github.com/teamcapybara/capybara) for our integration tests so we can interact with the app as a real user would.

Reviewer: OmbuLabs

This is the process we used to replaced `capybara-webit` gem in a legacy project with a more modern approach using the `webdrivers` gem and headless browser.

<!--more-->

## Issues With Capybara-webkit

- Development was officially suspended on March. [commit](https://github.com/thoughtbot/capybara-webkit/commit/f429d668568ff7349f5e23a085df7fcf1c431fa7#diff-04c6e90faac2675aa89e2176d2eec7d8)
- Depends on QT so it requires the developer to install extra libraries on the system adding an extra step to make a project run locally, inside docker, or in our CI system.
- It's difficult to update the webkit engine that's run in the background so you can't have all the new features you would have on the actual browsers

## Process

> We will devide this change into 4 different steps:

Reviewer: Divide

1. Replace `capybara-webkit`

> 1. Make sure specs passes locally

Reviewer: Isn't locally and docker the same in this case?

Author: Hmmm I meant just running `rake spec` using your OS ruby version and database and everything instead of inside the Docker container, maybe `natively`?

+ gem 'webdrivers'
```

This project was old so it also had an old capybara version that won't automatically register the browser drivers that we want so we had to register the headless browsers' drivers on our capybara config:

```ruby
# from https://github.com/teamcapybara/capybara/blob/c7c22789b7aaf6c1515bf6e68f00bfe074cf8fc1/lib/capybara/registrations/drivers.rb

Capybara.register_driver :headless_firefox do |app|
  Capybara::Selenium::Driver.load_selenium
  browser_options = ::Selenium::WebDriver::Firefox::Options.new
  browser_options.args << '-headless'
  Capybara::Selenium::Driver.new(app, browser: :firefox, options: browser_options)
end

Capybara.register_driver :headless_chrome do |app|
  Capybara::Selenium::Driver.load_selenium
  browser_options = ::Selenium::WebDriver::Chrome::Options.new.tap do |opts|
    opts.args << '--headless'
    opts.args << '--disable-gpu' if Gem.win_platform?
    # Workaround https://bugs.chromium.org/p/chromedriver/issues/detail?id=2650&q=load&sort=-id&colspec=ID%20Status%20Pri%20Owner%20Summary
    opts.args << '--disable-site-isolation-trials'
  end
  Capybara::Selenium::Driver.new(app, browser: :chrome, options: browser_options)
end
```

> > You'll need Chrome or Firefox versions that supports headless mode

Author: first one yes (it's just an extra comment about browser version), next one shouldn't be gray

And now we tell capybara to use one of those drivers for all the tests that require javascript:

```ruby
Capybara.javascript_driver = :headless_chrome # or :headless_firefox
```

> We had an extra step here because a lot of specs had the drivers specified on the actual test `describe` block like this: `describe "Do something", js: true, driver: :webkit do`. Instead of replacing the driver there, we just removed that option on all the tests so it uses the base config we set above.

And lastly, before you run your test, make you you have no references to capybara-webkit in your code (search for `capybara-webkit`, `Capybara::Webkit` and `:webkit` strings).

# Fixing Specs

> We first make sure tests are working locally so we are sure we don't have errors due to a misconfigured container.

Reviewer: What is a `misconfigured container` ?

Author: I meant a Docker container that may be missing some package or dependency that you have on your system and maybe never configured it on the Dockerfile (I ended up using Firefox instead of Chrome because Chrome threw me some errors, if wasn't sure the tests were working I may have thought that the problem was on the tests instead of the docker container). So to be sure that I'm not having errores because of a wrong docker file I first tried to be sure that the change actually worked without docker.

> This application uses an old version of [Chosen](https://harvesthq.github.io/chosen/) to customize the `select` tags. The test suite includes a helper method so the driver can select options from that custom select and that helper method started failing.

Reviewer: Just curious if this was related to the `capybara-webkit` change?

Author: This errors were only present on that branch after replacing the gem, and I think it makes sense since capybara-webkit was using an old deprecated QtWebView (which I think was bugged since I had another error where it wasn't actually hiding the original SELECT tag when replacing it with chosen and an expectation was succeeding when it should have failed)

> We were using a helper method from this [gist](https://gist.github.com/thijsc/1391107/699d65defed793eed0f04ead33c35737c641be53) that relied too much on `page.execute_script` and `page.evaluate_script`. The solution for this was to use a different method calling anly capybara methods to actually mimic the user interactions.

Reviewer: `only capybara methods`

```diff
  def select_from_chosen(item_text, options)
-    field = find('#' + options[:from], visible: false)
-    option_value = page.evaluate_script("$(\"##{field[:id]} option:contains('#{item_text}')\").val()")
-    page.execute_script("value = ['#{option_value}']\; if ($('##{field[:id]}').val()) {$.merge(value, $('##{field[:id]}').val())}")
-    option_value = page.evaluate_script("value")
-    page.execute_script("$('##{field[:id]}').val(#{option_value})")
-    page.execute_script("$('##{field[:id]}').trigger('liszt:updated').trigger('change')")

+    field = find_field(options[:from], visible: false)
+    find("##{field[:id]}_chzn").click
+    find("##{field[:id]}_chzn ul.chzn-results li", text: item_text).click
  end
```

> So we not only updated the gem, we also improved the test suit by not doing obscure Javascript calls.

Reviewer: `test suite`

# Updating Docker Container

Now that our tests are passing we can update the `Dockerfile` removing the commands that installed QT's libraries `qt5-default` and `libqt5webkit5-dev` (that will depend on the OS your are using, we have a Debian image so we just removed those packes from the `apt-get` command).

Then we make sure we have Mozilla Firefox or Google Chrome installed. It was easier to install Firefox since the `firefox-esr` package was already available on the Debian Buster image we were using. You can use a Docker image which includes other browsers or install Google Chrome if you want to.

So we changed the default driver to use Firefox:

```ruby

> Capybara.javascript_driver = :headless_firefox

Reviewer 2: if it was google chrome would you do `:headless_chrome` ?

Author: yes, the example using headless chrome is on the Replacing section after I have that notice to make sure the browsers are updated

```

Now our tests can be run inside Docker and we are all green.

# Updating CircleCI Config

> The last part of the process was to make sure the test passes on the CI system. We only needed to make sure Firefox was available when running the test, so the easiest solution for this was to use a CircleCI Ruby image and use the correct [variant](https://circleci.com/docs/2.0/circleci-images/#language-image-variants) so it also includes most used browsers:

Reviewer 2: ```suggestion
The last part of the process was to make sure the tests pass on the CI system. We only needed to make sure Firefox was available when running the test, so the easiest solution for this was to use a CircleCI Ruby image and use the correct [variant](https://circleci.com/docs/2.0/circleci-images/#language-image-variants) so it also includes most used browsers:
```

```diff
- - image: circleci/ruby:2.4.10-buster-node
+ - image: circleci/ruby:2.4.10-buster-node-browsers
```

> We did have some problems related to caching, so you may want to try to retry your CI job without caching just to make sure if you see something failing

## Conclusion

We finally got an old dependency removed (QT) replacing `capybara-webkit` with a more modern solution.

It makes it easier for new developers to join the project now that it has less extra requirements. It will also help when updating the Rails version, since newer Rails versions adds `webdrivers` gem when creating new apps by default.

> We also improved the test suite. When writing a feature spec (or system spec as introduced in Rails 5.2) you want to mimic the user interactions. If we don't test that the user can actually click an element and we trigger the `click` event using plain javascript, we may break something for the end user without noticing on the tests.

Reviewer 2: break something for the end user?

Author: hmm I meant that if we are not testing the user interactions using capybara methods (and just modify the state using javascript) we could introduce some code that breaks things that a user should be able to interact with and our tests won't detect that

Reviewer 2: maybe you just need to delete the word "end"?

---
layout: post

> title: "Removing capybara-webkit"

Reviewer 3: Technically is not removing, but replacing it...

So maybe: 

```suggestion
title: "How to Migrate from Capybara Webkit to Webdrivers"
```

date: 2020-04-15 12:00:00
categories: ["rails", "upgrades"]
author: arielj
---

We all know testing is important. We have our unit tests and integration tests to make sure everything is working.

At OmbuLabs, we use [capybara](https://github.com/teamcapybara/capybara) for our integration tests so that we can interact with the app as a real user would.

> Capybara by default uses `rack-test` as the driver, but `rack-test` does not support Javascript so to test some things we need a web browser with Javascript capabilities. Having the browser show up while the tests are running is not that practical, now we can use the headless feature on browsers like Chrome or Firefox (even Edge), but a few years ago the available options were mainly PhantomJS or capybara-webkit (through QtWebKit). Those options were deprecated in favor of the major browsers' headless mode.

Author: capybara-webkit development was suspended because of QtWebKit being deprecated https://github.com/thoughtbot/capybara-webkit#capybara-webkit (in favor of QtWebEngine though, not the major browsers).

I'm not sure if suspended because of qtwebkit being deprecated implies deprecation, so I'll use `development was suspended`

> Capybara by default uses `rack-test` as the driver, but `rack-test` does not support Javascript so to test some things we need a web browser with Javascript capabilities. Having the browser show up while the tests are running is not that practical, now we can use the headless feature on browsers like Chrome or Firefox (even Edge), but a few years ago the available options were mainly PhantomJS or capybara-webkit (through QtWebKit). Those options were deprecated in favor of the major browsers' headless mode.

Reviewer 3: @arielj Got it. Then that sentence would sound better like this: 

```
Capybara Webkit is no longer being developed in favor of ... 
```

This is the process we used to replace the `capybara-webkit` gem in a legacy project with the more modern approach of the `webdrivers` gem and headless browser.

> <!--more-->

Author: I moved the `Capybara by default...` part to a different section `Why Capybara Webkit?` after the `more` snippet

> ## Issues With Capybara-webkit

Author: Is it ok to modify the gem's name? they use `capybara-webkit` always

> ## Issues With Capybara-webkit

Reviewer 3: @arielj I think you should use "Capybara Webkit" or `capybara-webkit`? (not `Capybara-webkit`)

In their original announcement they used "Capybara Webkit" -> https://thoughtbot.com/blog/capybara-webkit

- Development was officially suspended in March. [commit](https://github.com/thoughtbot/capybara-webkit/commit/f429d668568ff7349f5e23a085df7fcf1c431fa7#diff-04c6e90faac2675aa89e2176d2eec7d8)
- Depends on QT so it requires the developer to install extra libraries on the system adding an extra step to make a project run locally, inside docker, or in our CI system.
- It's difficult to update the webkit engine that's run in the background so you can't have all the new features you would have on the actual browsers

## Process

We will divide this change into 4 different steps:
1. Replace `capybara-webkit`

> 1. Make sure specs passes locally

Reviewer 4: ```suggestion
1. Make sure specs pass locally
```

> 2. Make sure specs passes inside the Docker container

Reviewer 4: ```suggestion
2. Make sure specs pass inside the Docker container
```

> 3. Make sure specs passes in CircleCI

Reviewer 4: ```suggestion
3. Make sure specs pass in CircleCI
```

  Capybara::Selenium::Driver.new(app, browser: :firefox, options: browser_options)
end

Capybara.register_driver :headless_chrome do |app|
  Capybara::Selenium::Driver.load_selenium
  browser_options = ::Selenium::WebDriver::Chrome::Options.new.tap do |opts|
    opts.args << '--headless'
    opts.args << '--disable-gpu' if Gem.win_platform?
    # Workaround https://bugs.chromium.org/p/chromedriver/issues/detail?id=2650&q=load&sort=-id&colspec=ID%20Status%20Pri%20Owner%20Summary
    opts.args << '--disable-site-isolation-trials'
  end
  Capybara::Selenium::Driver.new(app, browser: :chrome, options: browser_options)
end
```
> You'll need Chrome or Firefox versions that supports headless mode

And now we tell capybara to use one of those drivers for all the tests that require javascript:

```ruby
Capybara.javascript_driver = :headless_chrome # or :headless_firefox
```

We had an extra step here because a lot of specs had the drivers specified on the actual test `describe` block like this: `describe "Do something", js: true, driver: :webkit do`. Instead of replacing the driver there, we just removed that option on all the tests so it uses the base config we set above.

> And lastly, before you run your test, make you you have no references to capybara-webkit in your code (search for `capybara-webkit`, `Capybara::Webkit` and `:webkit` strings).

Reviewer 4: ```suggestion
And lastly, before you run your test, make sure you have no references to capybara-webkit in your code (search for `capybara-webkit`, `Capybara::Webkit` and `:webkit` strings).
```

# Fixing Specs

> We first make sure tests are working locally so we are sure we don't have errors due to a misconfigured container.

Reviewer 4: ```suggestion
We first make sure tests are working locally so we don't have errors due to a misconfigured container.
```

> This application uses an old version of [Chosen](https://harvesthq.github.io/chosen/) to customize the `select` tags. The test suite includes a helper method so the driver can select options from that custom select and that helper method started failing.

Reviewer 4: ```suggestion
This application uses an old version of [Chosen](https://harvesthq.github.io/chosen/) to customize the `select` tags. The test suite includes a helper method so the driver can select options from that custom `select` and that helper method started failing.
```

We were using a helper method from this [gist](https://gist.github.com/thijsc/1391107/699d65defed793eed0f04ead33c35737c641be53) that relied too much on `page.execute_script` and `page.evaluate_script`. The solution for this was to use a different method calling only capybara methods to actually mimic the user interactions.

```diff
  def select_from_chosen(item_text, options)
-    field = find('#' + options[:from], visible: false)
-    option_value = page.evaluate_script("$(\"##{field[:id]} option:contains('#{item_text}')\").val()")
-    page.execute_script("value = ['#{option_value}']\; if ($('##{field[:id]}').val()) {$.merge(value, $('##{field[:id]}').val())}")
-    option_value = page.evaluate_script("value")
-    page.execute_script("$('##{field[:id]}').val(#{option_value})")
-    page.execute_script("$('##{field[:id]}').trigger('liszt:updated').trigger('change')")

+    field = find_field(options[:from], visible: false)
+    find("##{field[:id]}_chzn").click
+    find("##{field[:id]}_chzn ul.chzn-results li", text: item_text).click
  end
```

> So we not only updated the gem, we also improved the test suite by not doing obscure Javascript calls.

Reviewer 4: ```suggestion
So we not only updated the gem, we also improved the test suite by not doing obscure JavaScript calls.
```

# Updating Docker Container

Now that our tests are passing we can update the `Dockerfile` removing the commands that installed QT's libraries `qt5-default` and `libqt5webkit5-dev` (that will depend on the OS your are using, we have a Debian image so we just removed those packes from the `apt-get` command).

Then we make sure we have Mozilla Firefox or Google Chrome installed. It was easier to install Firefox since the `firefox-esr` package was already available on the Debian Buster image we were using. You can use a Docker image which includes other browsers or install Google Chrome if you want to.

So we changed the default driver to use Firefox:

```ruby
Capybara.javascript_driver = :headless_firefox
```

Now our tests can be run inside Docker and we are all green.

# Updating CircleCI Config

> The last part of the process was to make sure the tests pass on the CI system. We only needed to make sure Firefox was available when running the test, so the easiest solution for this was to use a CircleCI Ruby image and use the correct [variant](https://circleci.com/docs/2.0/circleci-images/#language-image-variants) so it also includes most used browsers:

Reviewer 4: ```suggestion
The last part of the process was to make sure the tests pass on the CI system. We only needed to make sure Firefox was available when running the test, so the easiest solution for this was to use a CircleCI's Ruby image and use the correct [variant](https://circleci.com/docs/2.0/circleci-images/#language-image-variants) so it also includes the most used browsers:
```

```diff
- - image: circleci/ruby:2.4.10-buster-node
+ - image: circleci/ruby:2.4.10-buster-node-browsers
```

> We did have some problems related to caching, so you may want to try to retry your CI job without caching just to make sure if you see something failing

## Conclusion

> We finally got an old dependency removed (QT) replacing `capybara-webkit` with a more modern solution.

Reviewer 3: ```suggestion
To sum up we got rid of an outdated dependency (`capybara-webkit`) which depended on QT and replaced it with `webdrivers` which depends on modern web browsers (e.g. Firefox). And we highly recommend you do it in your legacy Rails applications!
```

> It makes it easier for new developers to join the project now that it has less extra requirements. It will also help when updating the Rails version, since newer Rails versions adds `webdrivers` gem when creating new apps by default.

Reviewer 3: ```suggestion
It makes it easier to maintain our Rails application and it gets us closer to "the Rails way". It will help us when upgrading Rails because more recent versions of Rails rely on the `webdrivers` gem by default.
```

Reviewer 3: I don't think it makes it easier to onboard new developers, it's pretty much the same. It replaces capybara-webkit+qt with webdrivers+firefox. So _technically_ there are no fewer requirements...

Author: Maybe I didn't express that right. I mean that QT is a complicated dependency (https://github.com/thoughtbot/capybara-webkit/wiki/Installing-Qt-and-compiling-capybara-webkit) that you'll have to install just to run the tests suite, and firefox (or chrome) is usually there for any web developer at least for manual testing purposes, I'm not sure if that counts or not as a requirement.

Technically it may be the same amount of requirements, but I thought the whole point was to remove that qt dependency because it was harder to setup so I do think it helps to onboard new developers, maybe not because of fewer requirements but because of `easier` requirements?

What do you think?

Reviewer 3: @arielj Got it! Then maybe you can add a note about that: 

> Installing QT is quite complicated (https://github.com/thoughtbot/capybara-webkit/wiki/Installing-Qt-and-compiling-capybara-webkit). Firefox (or Chrome) are usually already installed in any web developer's environment.

Author: Thanks, I only changed the `Rails rely on the...` phrase, cause it gives the impression that you need that gem for newer Rails version but it's just the rails way and can be changed if someone wants to.

> Another advantage of replacing QtWebKit with a major browser like Chrome and Firefox is that it helped us to find a bug in our current test suite. We had not been properly testing the interactions of the user in those tests related to Chose, but now we fixed that.

Reviewer 4: ```suggestion
Another advantage of replacing QtWebKit with a major browser like Chrome and Firefox is that it helped us to find a bug in our current test suite. We had not been properly testing the interactions of the user in those tests related to Chosen, but now we fixed that.
```

> And finally, we can run the complete test suite using different browsers (`webdrivers` support Firefox, Chrome, IE and Edge) just changing one configuration if we need to.

Reviewer 4: ```suggestion
And finally, we can run the complete test suite using different browsers (`webdrivers` support Firefox, Chrome, IE and Edge) just by changing one configuration if we need to.
```
